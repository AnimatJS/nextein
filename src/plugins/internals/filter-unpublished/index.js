/**
 *
 * @param {Object} options
 * @param {String} options.field
 * @param {Array} posts
 */
export function prefilter ({ field = 'published' }, posts) {
  return posts.filter(p => p.data[field] !== false)
}
