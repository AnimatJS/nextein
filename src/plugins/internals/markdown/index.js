
import fm from 'frontmatter'
import parser from './parser'
import removePosition from 'unist-util-remove-position'

function normalizeDate (value, { path }) {
  if (!value) return

  try {
    return new Date(value)
  } catch (e) {
    console.warn(`Invalid date in frontmatter at: ${path}`)
  }
}

function parseContent (text, options) {
  const instance = parser(options)
  // second text argument avoids positoin information being removed in rehype-raw
  return instance.runSync(instance.parse(text), text)
}

function createOptions (source, raw) {
  const { data: extra = {}, content: text } = fm(raw)
  if (extra.date) extra.date = normalizeDate(extra.date, source)

  return {
    meta: {
      ...source,
      extra: {
        ...source.extra,
        ...extra
      }
    },
    content: text,
    raw
  }
}

/**
 * indexer
 * @param {Object} options
 * @param {Array} options.remark plugins for remark
 * @param {Array} options.rehype plugins for rehype
 * @param {Object} buildOptions
 * @param {Object} action
 * @param {Function} action.create
 */
export async function indexer (options, { load, ...source }, { create }) {
  if (source.mimeType === 'text/markdown') {
    const raw = await load()
    create(createOptions(source, raw, options))
  }
}

/**
 * build
 * @param {Object} options
 * @param {Array} options.remark plugins for remark
 * @param {Array} options.rehype plugins for rehype
 */
export function build (options, posts) {
  return posts.map(post => {
    if (post.data.mimeType === 'text/markdown') {
      post.content = parseContent(post.content, options)
    }
    return post
  })
}

/**
 * cleanup
 * @param {Object} options
 * @param {Boolean} options.raw set to true to keep raw value in post object. Default to false.
 * @param {Boolean} options.position set to true to keep position information in content hast. Default to false.
 * @param {Array} posts
 */
export function cleanup ({ raw = false, position = false }, posts) {
  return posts.map(post => {
    if (post.data.mimeType === 'text/markdown') {
      if (!position) post.content = removePosition(post.content, true /* hard delete */)
      if (!raw) delete post.raw
    }
    return post
  })
}
