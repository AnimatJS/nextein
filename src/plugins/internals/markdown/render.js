
import React from 'react'
import unified from 'unified'
import rehype from 'rehype-parse'
import stringify from 'rehype-stringify'
import reactRenderer from 'rehype-react'
import { select } from 'unist-util-select'

const extractExcerpt = (excerpt) => {
  const selector = (typeof excerpt === 'string') ? excerpt : ':root > element[tagName=p]'

  return () => /* attacher */ (tree) => {
    /* transformer */
    if (excerpt) {
      return select(selector, tree)
    }
    return tree
  }
}

export function render (_, { data, content, excerpt, renderers, components, prefix = 'entry-' }) {
  if (data.mimeType === 'text/markdown') {
    const hast = JSON.parse(JSON.stringify(content))
    const p = unified()
      .use(rehype)
      .use(extractExcerpt(excerpt))
      .use(stringify)
      .use(reactRenderer, {
        createElement: React.createElement,
        prefix,
        components: components || renderers
      })

    return p.stringify(p.runSync(hast))
  }
}
