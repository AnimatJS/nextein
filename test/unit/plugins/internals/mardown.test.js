
import { indexer, build } from  '../../../../src//plugins/internals/markdown'

const content = [
  '',
  'Text.'    
].join('\n')

const raw = [
  '---',
  'page: PAGE',
  'title: TITLE',
  '---',
  '',
  'Text.'    
].join('\n')

async function load () {
  return raw
}

describe('mardown::indexer', () => {

  test('default options', async () => {
    const options = {}
    const source = {
      mimeType: 'text/markdown'
    }

    expect.assertions(3);

    await indexer(options, { load, ...source }, { 
      create: function (entry) {
        expect(entry).toHaveProperty('meta')
        expect(entry).toHaveProperty('content')
        expect(entry).toHaveProperty('raw')
      }
    })
  })
})


describe('mardown::build', () => {

  test('default options', async () => {
    const options = {}
    const post = {
      data: { mimeType: 'text/markdown' },
      content,
      raw
    }

    const result = await build(options, [post])
    expect(result).toMatchSnapshot()
  })
})
