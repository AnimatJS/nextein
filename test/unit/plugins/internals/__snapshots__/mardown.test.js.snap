// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mardown::build default options 1`] = `
Array [
  Object {
    "content": Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "position": Object {
                "end": Object {
                  "column": 6,
                  "line": 2,
                  "offset": 6,
                },
                "start": Object {
                  "column": 1,
                  "line": 2,
                  "offset": 1,
                },
              },
              "type": "text",
              "value": "Text.",
            },
          ],
          "position": Object {
            "end": Object {
              "column": 6,
              "line": 2,
              "offset": 6,
            },
            "start": Object {
              "column": 1,
              "line": 2,
              "offset": 1,
            },
          },
          "properties": Object {},
          "tagName": "p",
          "type": "element",
        },
      ],
      "data": Object {
        "quirksMode": false,
      },
      "position": Object {
        "end": Object {
          "column": 6,
          "line": 2,
          "offset": 6,
        },
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "root",
    },
    "data": Object {
      "mimeType": "text/markdown",
    },
    "raw": "---
page: PAGE
title: TITLE
---

Text.",
  },
]
`;
